libgadget/fof.c

```diff
@@ -442,33 +424,38 @@ fofp_merge(int target, int other, TreeWalk * tw)
 {
     /* this will lock h1 */
     int * Head = FOF_PRIMARY_GET_PRIV(tw)->Head;
-    struct SpinLocks * spin = FOF_PRIMARY_GET_PRIV(tw)->spin;
     int h1, h2;
-
     do {
-        h1 = HEADl(-1, target, -1, Head, spin);
-        /* stop looking if we find h1 along the path (because it is already owned by us) */
-        h2 = HEADl(h1, other, h1, Head, spin);
-        /* We had a lock already taken on h2 by another thread.
-         * We need to unlock h1 and retry to avoid deadlock loops.*/
-        if(h2 == -2)
-            unlock_spinlock(h1, spin);
-    } while(h2 == -2);
-
-    if(h2 >=0)
-    {
-        /* h2 as a sub-tree of h1 */
-        Head[h2] = h1;
-
-        /* update MinID of h1 */
-        if(HaloLabel[h1].MinID > HaloLabel[h2].MinID)
-        {
-            HaloLabel[h1].MinID = HaloLabel[h2].MinID;
-            HaloLabel[h1].MinIDTask = HaloLabel[h2].MinIDTask;
+        h1 = HEADl(-1, target, Head);
+        /* Done if we find h1 along the path
+         * (because other is already in the same halo) */
+        h2 = HEADl(h1, other, Head);
+        if(h2 < 0)
+            return;
+        /* Ensure that we always merge to the lower entry*/
+        if(h1 > h2) {
+            int tmp = h2;
+            h2 = h1;
+            h1 = tmp;
         }
-        //printf("unlocking %d by %d in merge\n", h2, omp_get_thread_num());
-        unlock_spinlock(h2, spin);
+     /* Atomic compare exchange to make h2 a subtree of h1.
+      * Set Head[h2] = h1 iff Head[h2] is still h2. Otherwise loop.*/
+    } while(!__atomic_compare_exchange(&Head[h2], &h2, &h1, 0, __ATOMIC_RELAXED, __ATOMIC_RELAXED));
+
+    struct SpinLocks * spin = FOF_PRIMARY_GET_PRIV(tw)->spin;
+
+    /* update MinID of h1: h2 is now just another child of h1
+     * so h2 does not need to be locked.
+     * h1 may not be its own head anymore,
+     * so we need to double check later.
```

Comment:
If we need to double check later, then perhaps the update can be entirely deferred? I feel this is possible as the loop above is no longer touching the MinID and MinTask.

Reply:
If it cannot be entirely deferred (e.g. interaction with the remote particles on mode==1), then we must be relying on some less level of correctness here? Would the catch later be still correct for the remote particles?

Reply:
The update can be entirely deferred, indeed. So we would lose all the locking in fofp_merge. 
Both the current code and the catch later are correct, I think, but I am not sure which performs better.

The issue is exactly interaction with the remote particles: I am concerned that we would end up with a lot more updates from the remote and so more contention. I might run a test on a large sized problem.

Reply:
I wanted to use atomic or compare-exchange on MinID/MinIDTask but it needs to be a scalar variable.

Reply:
It does seem a little slower if I entirely defer the update loop, although this is a small problem and the difference is only a few percent. I think I might leave that to a future PR, maybe to use thread-locals in fof and remove the spinlocks completely.

Reply:
There is a DWCAS on x86 that gcc seems to support (128 bit). Declaring the type as int128_t could probably work out of box. 

I compiled the code at:
https://gcc.gnu.org/legacy-ml/gcc-help/2017-06/msg00156.html

and it seems to be reasonable, calling "__atomic_compare_exchange_16@LIBATOMIC_1.0".

Reply:
Brilliant idea! I see also with godbolt that it works for icc > 17 and clang as well. I will implement this. Is the aligned attribute necessary?

Reply:
Alas, this makes it slower and more unreliable.